### PR TITLE
Fix issue #26610

### LINKED ISSUE #26610: QuickInfo displays non-generic XML comments when highlighting completion item is a generic type
**Version Used**: 
Visual Studio 2017 (27702.1.d15.8)

**Steps to Reproduce**:

1. Create a C# project
2. In the Main method, type `ImmutableArray`, or any identifier that has both non-generic and generic types (for example, `ValueTuple`)
3. Use the Down Arrow or mouse click to select the generic type in the IntelliSense items
4. A tooltip with type definition and XML comments (if any) will be displayed

**Expected Behavior**:
When non-generic type is highlighted, it should display XML comments from non-generic type declaration, when generic type is highlighted, it should display XML comments from generic type declaration.

**Actual Behavior**:
It displays non-generic type XML comments (the first match with the identifier,) for both non-generic and generic type definitions. See screenshot below.

image

### PR DESCRIPTION
**Pull request by mazhou-msft**
Fixes #26610

The PR follows the approved behaviors (https://github.com/dotnet/roslyn/issues/26610#issuecomment-446211056) mentioned in the issue discussion. 

Root cause: When constructing completion item description, we get symbol name and kind from the completion item property bags, and then get candidate symbols from the position. 

Say we have following types:

`System.ValueTuple`
`System.ValueTuple<T>`

Both types will have same symbol name and kind, hence when get candidate symbols for the two types, the symbols are the same.

The simple fix is to add `IsGeneric` to the `CompleteItem.Properties` bag, so that when loading candidate symbols for completion description, we can differ non-generic/generic symbols and give accurate QuickInfo texts.

New behavior:

image
image

### CODE REVIEW

File: src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs [hunk 1]
```diff
@@ -122,15 +122,31 @@ private static int ComputeSymbolMatchPriority(ISymbol symbol)
...
             var name = SymbolCompletionItem.GetSymbolName(item);
             var kind = SymbolCompletionItem.GetKind(item);
+            bool.TryParse(SymbolCompletionItem.GetSymbolIsGeneric(item), out var isGeneric);
             var relatedDocumentIds = document.Project.Solution.GetRelatedDocumentIds(document.Id).Concat(document.Id);
             var options = document.Project.Solution.Workspace.Options;
             var totalSymbols = await base.GetPerContextSymbols(document, position, options, relatedDocumentIds, preselect: false, cancellationToken: cancellationToken).ConfigureAwait(false);
             foreach (var (documentId, syntaxContext, symbols) in totalSymbols)
             {
-                var bestSymbols = symbols.Where(s => kind != null && s.Kind == kind && s.Name == name).ToImmutableArray();
+                var bestSymbols = symbols.Where(s => kind != null && s.Kind == kind && s.Name == name);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs]:
```suggestion
                var bestSymbols = symbols.Where(s => kind != null && s.Kind == kind && s.Name == name && isGeneric == (s.GetArity() != 0));
```

File: src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs [hunk 2]
```diff
@@ -122,15 +122,31 @@ private static int ComputeSymbolMatchPriority(ISymbol symbol)
...
             var name = SymbolCompletionItem.GetSymbolName(item);
             var kind = SymbolCompletionItem.GetKind(item);
+            bool.TryParse(SymbolCompletionItem.GetSymbolIsGeneric(item), out var isGeneric);
             var relatedDocumentIds = document.Project.Solution.GetRelatedDocumentIds(document.Id).Concat(document.Id);
             var options = document.Project.Solution.Workspace.Options;
             var totalSymbols = await base.GetPerContextSymbols(document, position, options, relatedDocumentIds, preselect: false, cancellationToken: cancellationToken).ConfigureAwait(false);
             foreach (var (documentId, syntaxContext, symbols) in totalSymbols)
             {
-                var bestSymbols = symbols.Where(s => kind != null && s.Kind == kind && s.Name == name).ToImmutableArray();
+                var bestSymbols = symbols.Where(s => kind != null && s.Kind == kind && s.Name == name && isGeneric == (s.GetArity() != 0));
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs]: oops, sorry, didn't mean to commit to your branch directly.  that was supposed to be a comment.

File: src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs [hunk 3]
```diff
@@ -122,15 +122,31 @@ private static int ComputeSymbolMatchPriority(ISymbol symbol)
...
+                        bestSymbols = isGeneric
+                            ? bestSymbols.Where(s => ((IMethodSymbol)s).Arity != 0)
+                            : bestSymbols.Where(s => ((IMethodSymbol)s).Arity == 0);
+                        break;
+                    case SymbolKind.NamedType:
+                        bestSymbols = isGeneric
+                            ? bestSymbols.Where(s => ((INamedTypeSymbol)s).Arity != 0)
+                            : bestSymbols.Where(s => ((INamedTypeSymbol)s).Arity == 0);
+                        break;
+                }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs]:
```suggestion
```

File: src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs]:
```suggestion
                var bestSymbols = symbols.WhereAsImmutable(s => kind != null && s.Kind == kind && s.Name == name && isGeneric == (s.GetArity() != 0));
```

File: src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs [hunk 4]
```diff
@@ -59,11 +59,19 @@ public static CompletionItem AddSymbolEncoding(IReadOnlyList<ISymbol> symbols, C
...
-        public static CompletionItem AddSymbolNameAndKind(IReadOnlyList<ISymbol> symbols, CompletionItem item)
+        public static CompletionItem AddSymbolInfo(IReadOnlyList<ISymbol> symbols, CompletionItem item)
         {
             var symbol = symbols[0];
+            var isGeneric = symbol switch
+            {
+                IMethodSymbol methodSymbol => methodSymbol.Arity != 0,
+                INamedTypeSymbol typeSymbol => typeSymbol.Arity != 0,
+                _ => false
+            };
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs]:
```suggestion. 
            var isGeneric = symbol.GetArity() > 0;
```

File: src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs [hunk 5]
```diff
@@ -326,6 +334,16 @@ internal static string GetSymbolName(CompletionItem item)
...
 
+        internal static string GetSymbolIsGeneric(CompletionItem item)
+        {
+            if (item.Properties.TryGetValue("IsGeneric", out var name))
+            {
+                return name;
+            }
+
+            return null;
+        }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs]:
```suggestion
        => item.Properties.TryGetValue("IsGeneric", out var name) ? name : null;
```

File: src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs (hunk 2 above)
Comment (mazhou-msft) [reply src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs]: I was not aware there is a `GetArity()` extension method, fixed. Thanks for the suggestions!

**Comment by CyrusNajmabadi** [comment]
@genlu Do we have a way to test this out to make sure it doesn't regress any completion performance?

File: src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs [hunk 6]
```diff
@@ -59,11 +59,14 @@ public static CompletionItem AddSymbolEncoding(IReadOnlyList<ISymbol> symbols, C
...
-        public static CompletionItem AddSymbolNameAndKind(IReadOnlyList<ISymbol> symbols, CompletionItem item)
+        public static CompletionItem AddSymbolInfo(IReadOnlyList<ISymbol> symbols, CompletionItem item)
         {
             var symbol = symbols[0];
+            var isGeneric = symbol.GetArity() > 0;
+
             return item.AddProperty("SymbolKind", ((int)symbol.Kind).ToString())
-                       .AddProperty("SymbolName", symbol.Name);
+                       .AddProperty("SymbolName", symbol.Name)
+                       .AddProperty("IsGeneric", isGeneric.ToString());
```
Comment (genlu) [inline src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs]: Can we just not add `IsGeneric` when it's not a generic type, and use the existence of the property as the indication instead? This way we would avoid creating a new item for majority of the items that are not generic.

**Comment by genlu** [comment]
@CyrusNajmabadi No we don't have perf tests for completion. Given this only affect symbol completion provider, not import completion, I'm not too concerned about it.

File: src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs (hunk 6 above)
Comment (mazhou-msft) [reply src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs]: Sure. Fixed.

File: src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs [hunk 7]
```diff
@@ -302,29 +307,18 @@ public static string GetInsertionText(CompletionItem item)
...
-            if (item.Properties.TryGetValue("SymbolKind", out var kind))
-            {
-                return (SymbolKind)int.Parse(kind);
-            }
+            => item.Properties.TryGetValue("SymbolKind", out var kind) ? (SymbolKind?)int.Parse(kind) : null;
 
-            return null;
-        }
+        internal static string GetSymbolIsGeneric(CompletionItem item)
+            => item.Properties.TryGetValue("IsGeneric", out var isGeneric) ? isGeneric : null;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs]:
```suggestion
        internal static bool GetSymbolIsGeneric(CompletionItem item)
            => item.Properties.TryGetValue("IsGeneric", out var v) && bool.TryParse(v, out var isGeneric) && isGeneric;
```

File: src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs [hunk 8]
```diff
@@ -122,12 +122,14 @@ private static int ComputeSymbolMatchPriority(ISymbol symbol)
             var position = SymbolCompletionItem.GetContextPosition(item);
             var name = SymbolCompletionItem.GetSymbolName(item);
             var kind = SymbolCompletionItem.GetKind(item);
+            bool.TryParse(SymbolCompletionItem.GetSymbolIsGeneric(item), out var isGeneric);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs]:
```suggestion
            var isGeneric = SymbolCompletionItem.GetSymbolIsGeneric(item);
```

File: src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs [hunk 9]
```diff
@@ -122,12 +122,14 @@ private static int ComputeSymbolMatchPriority(ISymbol symbol)
...
             var name = SymbolCompletionItem.GetSymbolName(item);
             var kind = SymbolCompletionItem.GetKind(item);
+            var isGeneric = SymbolCompletionItem.GetSymbolIsGeneric(item);
             var relatedDocumentIds = document.Project.Solution.GetRelatedDocumentIds(document.Id).Concat(document.Id);
             var options = document.Project.Solution.Workspace.Options;
             var totalSymbols = await base.GetPerContextSymbolsAsync(document, position, options, relatedDocumentIds, preselect: false, cancellationToken: cancellationToken).ConfigureAwait(false);
             foreach (var (documentId, syntaxContext, symbols) in totalSymbols)
             {
-                var bestSymbols = symbols.Where(s => kind != null && s.Kind == kind && s.Name == name).ToImmutableArray();
+                var bestSymbols = symbols.WhereAsArray(s => kind != null && s.Kind == kind && s.Name == name && isGeneric == (s.GetArity() > 0));
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs]:
```suggestion
                var bestSymbols = symbols.WhereAsArray(
                    s => kind != null && s.Kind == kind && s.Name == name && isGeneric == (s.GetArity() > 0));
```

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**